t/rest_server/main.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package main implements a simple HTTP/JSON REST example.
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"net/http"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/abcxyz/pkg/logging"
+	"github.com/abcxyz/pkg/renderer"
+	"github.com/abcxyz/pkg/serving"
+	"github.com/go-chi/chi/v5"
+)
+
+const defaultPort = "8080"
+
+var port = flag.String("port", defaultPort, "Specifies server port to listen on.")
+
+func handleHello(h *renderer.Renderer) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		logger := logging.FromContext(r.Context())
+		logger.Info("handling request")
+		h.RenderJSON(w, http.StatusOK, map[string]string{"message": "hello world"})
+	})
+}
+
+// RealMain creates an example backend HTTP server.
+// This server supports graceful stopping and cancellation.
+func realMain(ctx context.Context) error {
+	logger := logging.FromContext(ctx)
+	// Make a new renderer for rendering json.
+	// Don't provide filesystem as we don't have templates to render.
+	h, err := renderer.New(ctx, nil,
+		renderer.WithOnError(func(err error) {
+			logger.Errorw("failed to render", "error", err)
+		}))
+	if err != nil {
+		return fmt.Errorf("failed to create renderer for main server: %w", err)
+	}
+
+	r := chi.NewRouter()
+	r.Mount("/", handleHello(h))
+	walkFunc := func(method, route string, handler http.Handler, middlewares ...func(http.Handler) http.Handler) error {
+		logger.Debugw("Route registered", "http_method", method, "route", route)
+		return nil
+	}
+
+	if err := chi.Walk(r, walkFunc); err != nil {
+		logger.Errorw("error walking routes", "error", err)
+	}
+
+	httpServer := &http.Server{
+		Addr:              *port,
+		Handler:           r,
+		ReadHeaderTimeout: 2 * time.Second,
+	}
+
+	logger.Info("starting server on ", *port)
+	server, err := serving.New(*port)
+	if err != nil {
+		return fmt.Errorf("error creating server: %w", err)
+	}
+
+	// This will block until the provided context is cancelled.
+	if err := server.StartHTTP(ctx, httpServer); err != nil {
+		return fmt.Errorf("error starting server: %w", err)
+	}
+	return nil
+}
+
+func main() {
+	ctx, done := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer done()
+	logger := logging.FromContext(ctx)
+
+	flag.Parse()
+	if err := realMain(logging.WithLogger(ctx, logger)); err != nil {
+		done()
```

Comment:
Maybe add a comment here explaining that the call to done() is made because deferred function calls won't execute due to logger.Fatal() never returning.

Reply:
Added comment. Do the t.Fatals in the test file also need done() calls in the error handling if?

Reply:
Discussed offline, t.Fatals don't require this.